### ISSUE TITLE
Package Security Check

### ISSUE THREAD

**Issue by mitchellsimoens** [algolia/npm-search#72]
I think it'd be quite helpful to check the package and it's dependencies for security issues like [nsp](https://nodesecurity.io/) and display it in the package details page and may even be a great idea to put in the search result list.

**Comment by Haroenv** [comment]
That's a great idea, it could be added as another source here, by batching the calls if possible, it would be similar to how the downloads are fetched. Do you want to work on this @mitchellsimoens?

**Comment by rdela** [comment]
@mitchellsimoens @Haroenv I really like this idea.
Anything that helps people only install secure packages is a big win to me. Plus I need to get a feel for the code and what is possible with query rules anyway. Happy to give it a try if one of you willing to supervise me and review. What do I need to do to get setup with the right kind of token to test changes locally?

Maybe someone from @nodesecurity would be willing to let us know whether they are on board and possibly review final interface and/or implementation?

**Comment by MartinKolarik** [comment]
👍 we previously considered showing an "insecure" badge in the search results at https://www.jsdelivr.com and having this information available directly in the index would definitely make it easier.

**Comment by Haroenv** [comment]
You can request the data from 200 packages at the same time, I have a small demo I can push where other people can help

**Comment by rdela** [comment]
@Haroenv if you already started something I would love to help on that branch, maybe you can split up work and assign me little tasks I can attempt to address in single commits?

**Comment by Haroenv** [comment]
https://github.com/algolia/npm-search/tree/feat/nsp, but it's pretty late here and there's some bug that needs fixing where the `security` key doesn't show up in the relevant packages

**Comment by Haroenv** [comment]
You need an Algolia app with api key and index. If you tell me which appid  I can give you more records and operations so you don’t need to pay  @rdela

**Comment by rdela** [comment]
Nice. Signed in with GH earlier and went through onboarding, it was stellar and docs look like they are same caliber.

My first round of questions:

Ok to use this appid made during tour? If so, ok to post that here in issue or should I email/DM it to you somewhere?

Tried to create new app and got…
>  You cannot create new applications during the trial period

I guess the default appId (_`OFCNCOG2CU`_) [listed in the readme](https://github.com/algolia/npm-search#env-variables) is the Yarn version?

Looks like I can just [`clearIndex`/`deleteIndex`](https://www.algolia.com/doc/api-client/manage-indices/) before each test?

Do I need to worry about `indexName` or ok to leave off and 
`seq=0 apiKey=... yarn dev`
will just create _`npm-search`_ each time if it does not already exist?

Assuming readme [Usage > Development](https://github.com/algolia/npm-search#development) production warning…
> Be careful to develop on a different index than the production one when necessary.

 will not apply because I will not have a “production” index yet?

**Comment by Haroenv** [comment]
> Nice. Signed in with GH earlier and went through onboarding, it was stellar and docs look like they are same caliber.
> 
> My first round of questions:
> 
> Ok to use this appid made during tour? If so, ok to post that here in issue or should I email/DM it to you somewhere?

Yes that should be good. The appId is not exactly secret so you can send it here if you need more usage
 
> Tried to create new app and got…
> 
> You cannot create new applications during the trial period

oh yeah that's right, I think if you pick a (free) plan you can create new apps, just not in trial. 

> I guess the default appId (OFCNCOG2CU) listed in the readme is the Yarn version?

Yes, that's the appid used in production

> Looks like I can just clearIndex/deleteIndex before each test?

yep, both are equivalent for this usage.
>
> Do I need to worry about indexName or ok to leave off and
> seq=0 apiKey=... yarn dev

you can put the index name in config.js (where it has a default of `npm-search`)

> will just create npm-search each time if it does not already exist?

it will create or append to that index

> Assuming readme Usage > Development production warning…
>
> Be careful to develop on a different index than the production one when necessary.
> will not apply because I will not have a “production” index yet?

This is for people at Algolia with access to the production index.

**Comment by rdela** [comment]
>  The appId is not exactly secret so you can send it here if you need more usage

Cool, appid:
ERNLE6UZ7H

Trial seems pretty generous and I am only testing this for now. If you want to wait to uncap til I hit limits or trial expires in 13 days, fine by me, can mention you here then and include key again.

Any reason to get rid of `getstarted_actors` demo index before I start testing? Won’t bother testing `npm-search` index somehow will it?

Like the Actors example and way it demos searchable attributes and custom ranking formula, might refer to it later on, but could always create another account and leave in free plan for that.

So would you mind elaborating on:
> `security` key doesn't show up in the relevant packages

…the one we `map` over in `nsp.js`? Where is it missing? In the index? I have a hunch about introduction of `securityRecommendation` in [chore: spacing and less reduncancy](https://github.com/algolia/npm-search/commit/bc6b8894f2fe76006c8f7c7a11850eb3f4e55be9#diff-10a95e0252ed05e7f319b65ffe7525c3R73), which does not seem to occur elsewhere?
Should that be plain [`recommendation`](https://github.com/algolia/npm-search/blob/feat/nsp/nsp.js#L73)?

Want to make sure I know how to tell if/when it is fixed before I dive in though.

**Comment by Haroenv** [comment]
I think I did something wrong when mapping in the weekend, the data wasn't present in the index, I haven't taken a look at what the cause was yet. It'll be fixed if the records have a `securityRecommendation` or whatever it's renamed to in the index. 

You can remove or keep the getting started index. 

Let us know if you want extended records/operations anytime

**Comment by rdela** [comment]
Great news all around, think I am headed in the right direction, will try to PR in next day or so, thanks Haroen!

**Comment by rdela** [comment]
@Haroenv thanks for stopping by the Gatsby issue, this is still on my radar and I was not trying to bring this discussion over there, only to ask you about Gatsby plugin index design in progress in case you had ideas from all of your experience implementing and helping others implement search. I am not sure whether we will end up displaying security info in the initial Gatsby index yet even though I personally am heavily in favor.

Part of the reason I mentioned you was the idea to show some sort of indication that a package is not 1.0 yet is interesting, and may be worth me opening a separate issue on the Yarn site repo about. But we can circle back to it once this is merged.

Will let you know how I get on with the code tomorrow and over the weekend, have some time set aside for this so should be able to have something by Monday. Will you be the one implementing the Yarn interface? Is it worth trying to get it displaying in their templates in some form once I get data matched correctly or we would need to work that out later?

**Comment by Haroenv** [comment]
I or someone else can do it on yarnpkg/website. It should be pretty straightforward to simply display it a little like the deprecated view and some link back to nsp @rdela

**Comment by rdela** [comment]
Cool thanks @Haroenv will PR this before I mess with Yarn site

**Comment by rdela** [comment]
@Haroenv looking now, thanks for https://github.com/algolia/npm-search/commit/0a2e39ede60dc355b82c765fa8fc61bca84c872d

**Comment by rdela** [comment]
@Haroenv Ok so I am even with you now on my fork, first run fails with:
[`AlgoliaSearchError: Record at the position 73 objectID=101 is too big size=23909 bytes. Contact us if you need an extended quota`](https://gist.github.com/rdela/ab40ff141de050d44f85db084955ca78#file-npm-search-feat-nsp-20171204-144227-txt-L5)

Is this a me still being in trial issue?

We are pulling [`securityRecommendation` data](https://gist.github.com/rdela/ab40ff141de050d44f85db084955ca78#file-npm-search-feat-nsp-20171204-144227-txt-L14) though

**Comment by rdela** [comment]
@Haroenv [confirmed in explorer](https://gist.github.com/rdela/ab40ff141de050d44f85db084955ca78#gistcomment-2275943)

**Comment by rdela** [comment]
@Haroenv added both of your emails I saw glancing through commit logs as collaborators on my trial app so you can see firsthand if you like

**Comment by Haroenv** [comment]
@rdela, yes there will be some records that are too big for a regular index, you can filter those out. Is it the nsp data which is too big? Maybe we should try to find a more data conservative way of writing it 🤔

**Comment by jimaek** [comment]
Is this being worked on?

**Comment by Haroenv** [comment]
Feel free to work on this @jimaek

**Comment by rdela** [comment]
Yes, exactly what Haroen said, feel free to work on this. I have not touched it since comments above from Dec. 2017 but am willing to try to help anyone make progress if they want to attempt to bring it up to date. That said, my time is very limited currently so I may not be of much assistance.

The [feat/nsp branch](https://github.com/rdela/npm-search/tree/feat/nsp) Haroen created is still available on my fork even though it has been deleted from this repo.

NSP/^Lift were [swallowed by npm](https://blog.npmjs.org/post/172793182214/npm-acquires-lift-security-and-node-security) in April 2018. The [nsp package](https://www.npmjs.com/package/nsp) was deprecated 7 May 2018 and the 
[API shut down 30 Sept 2018](https://blog.npmjs.org/post/175511531085/the-node-security-platform-service-is-shutting).

Although the use case is different since it wants a `package.json`, looks like there may be some prior art to pull ideas from in the [npm-auditor](https://www.npmjs.com/package/npm-auditor) package ([Github](https://github.com/AndreyBelym/npm-auditor)) that uses [npm-registry-fetch](https://www.npmjs.com/package/npm-registry-fetch) and [npm-audit-report](https://www.npmjs.com/package/npm-audit-report).

**Comment by jimaek** [comment]
Ok, we will check it out since we want to use this at jsDelivr

**Comment by bodinsamuel** [comment]
@jimaek do you still need it? 
I see that you have a widget in jsdelivr now, so probably not a strong push anymore. 
Could be interesting to have it in the index directly though, if any api exists

**Comment by jimaek** [comment]
In our case Snyk sent a PR directly https://github.com/jsdelivr/www.jsdelivr.com/pull/330
I am certain they would be glad for you to integrate their data into your index. So maybe contact Brian to make sure you don't overload their widget :)
